File: problem_14/problem_14.py
import time

def check_chain_length(n):
    if n == 1:
        return 1
    if n in chain_table:
        return chain_table[n]
    if n % 2 == 0:
        length = check_chain_length(n//2) + 1
    else: 
        length = check_chain_length((3*n + 1)//2)+ 2
    if n < 1000000:
        chain_table[n] = length
    return length

def problem_14():
    longest_chain, number = 1,1
    for n in range(500000,1000000):
        chain_length = check_chain_length(n)
        if chain_length > longest_chain:
            longest_chain,number = chain_length, n
    return number

start = time.time()
chain_table = {}
solution = problem_14()
print(f"{solution} found in {time.time() - start} seconds.")
# 837799 found in 0.7883999347686768 seconds.

File: problem_14/problem_14.js
const now = require("performance-now");

const checkChainLength = n => {
    var length;
    if (n === 1){
        return 1;
    }
    if (chainTable[n]){
        return chainTable[n]
    }
    if (n % 2 === 0){
        length = checkChainLength( Math.floor(n / 2) ) + 1
    }
    else {
        length = checkChainLength( Math.floor( (3*n + 1) / 2 )) + 2
    }
    if ( n < 1000000){
        chainTable[n] = length
    }
    return length
}

const problem14 = () => {
    var longestChain = 1;
    var number = 1;
    var chainLength;
    for (var n = 500000; n < 1000000; n++){
        chainLength = checkChainLength(n);
        if (chainLength > longestChain){
            longestChain = chainLength;
            number = n;
        }
    }
    return number;
}

let start = now();
const chainTable = {};
let solution = problem14()
console.log(`${solution} found in ${(now()-start) / 1000} seconds.`)
// 837799 found in 0.214156243 seconds.